### ISSUE TITLE
UndefinedError: 'str object' has no attribute 'meta'

### ISSUE THREAD

**Issue by zbeekman** [cmacmackin/ford#139]
Is there a down side to including more verbose debug output by default? I'm encountering a strange error with JSON-Fortran and 4.6.0  that did not exist in the previous release, and I only encounter it in the homebrew build sandbox environment...

Are there performance penalties (or other penalties) that  arise if you always provide more detailed debugging information?

**Comment by cmacmackin** [comment]
Without the `--debug` flag, FORD will just skip a file if it encounters 
an error while processing it. It also gives tidier looking error 
messages (no backtrace).

On 16-05-31 06:10 PM, Izaak Beekman wrote:
>
> Is there a down side to including more verbose debug output by 
> default? I'm encountering a strange error with JSON-Fortran and 4.6.0 
> that did not exist in the previous release, and I only encounter it in 
> the homebrew build sandbox environment...
>
> Are there performance penalties (or other penalties) that arise if you 
> always provide more detailed debugging information?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/139>, or mute the thread 
> <https://github.com/notifications/unsubscribe/AHxJPTztwpCdoIo_Ol4cE5JDnd8pE4Z6ks5qHGuTgaJpZM4IqwMY>.
>

-- 
Chris MacMackin
cmacmackin.github.io <http://cmacmackin.github.io>

**Comment by zbeekman** [comment]
Huh, ok. Im not sure how I feel about the "just skip a file" aspect. Does it at least emit a warning in this case?

Here is the context of my my question:
I am submitting Pull Requests to Homebrew to update:
GCC => 6.1.0
JSON-Fortran => 5.0.0
FORD => 4.6.0

However something very strange is happening in the Homebrew build environment/sandbox. After upgrading FORD, JSON-Fortran (past and present versions) cause FORD to crap out, and since by default we don't pass `--debug` (which based on your description, I think I will always pass `--debug` from now on) the Homebrew build is failing.

GCC 6.1.0 contains a regression with strings and GFortran, and JSON-Fortran 5.0.0 contains a work around for that regression. As such,  updating JSON-Fortran to 5.0.0 in Homebrew is a prerequisite to updating GCC to 6.1.0. This means, that until I can figure out what is happening in the Homebrew environment, we will have to wait on merging the FORD 4.6.0  PR and getting it into Homebrew/Linuxbrew.

**Comment by cmacmackin** [comment]
Yes, a warning is emitted when a bad file is encountered. Is this error 
you're getting with FORD the one Jacob Williams recently opened an issue 
about?

On 16-05-31 10:08 PM, Izaak Beekman wrote:
>
> Huh, ok. Im not sure how I feel about the "just skip a file" aspect. 
> Does it at least emit a warning in this case?
>
> Here is the context of my my question:
> I am submitting Pull Requests to Homebrew to update:
> GCC => 6.1.0
> JSON-Fortran => 5.0.0
> FORD => 4.6.0
>
> However something very strange is happening in the Homebrew build 
> environment/sandbox. After upgrading FORD, JSON-Fortran (past and 
> present versions) cause FORD to crap out, and since by default we 
> don't pass |--debug| (which based on your description, I think I will 
> always pass |--debug| from now on) the Homebrew build is failing.
>
> GCC 6.1.0 contains a regression with strings and GFortran, and 
> JSON-Fortran 5.0.0 contains a work around for that regression. As 
> such, updating JSON-Fortran to 5.0.0 in Homebrew is a prerequisite to 
> updating GCC to 6.1.0. This means, that until I can figure out what is 
> happening in the Homebrew environment, we will have to wait on merging 
> the FORD 4.6.0 PR and getting it into Homebrew/Linuxbrew.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/139#issuecomment-222821215>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe/AHxJPZ9TbTjd9d1th-IcH5doVgO9zhctks5qHKNZgaJpZM4IqwMY>.
>

-- 
Chris MacMackin
cmacmackin.github.io <http://cmacmackin.github.io>

**Comment by jacobwilliams** [comment]
#137 was for a different project (I don't think it should not be affecting JSON-Fortran).

**Comment by zbeekman** [comment]
@cmacmackin I don't think so. I'm not sure which you're referring too, perhaps #138? This error goes away when I build JSON-Fortran using FORD 4.6.0 in my environment, but the sanitized Homebrew environment does something that just causes FORD to not create any output and encounter an error. When I build JSON-Fortran using the same cmake with the same settings outside of the Homebrew build sandbox I don't have this issue. So I may have done something dumb in the FORD formula, or there might be something subtle happening with FORD in the sandbox environment.

**Comment by cmacmackin** [comment]
Yes, I was referring to #138. I don't know anything about the Homebrew sandbox, so it's difficult for me to be able to help. Could you post the error message (with backtrace), though? I might be able to narrow things down.

**Comment by zbeekman** [comment]
yes its a bit tricky to get the backtrace since, at the moment it pulls in a released version of JSON-Fortran, which has no easy/convenient means of adding the --debug flag while attempting a Homebrew build... that's what I'm working on now.

**Comment by cmacmackin** [comment]
Don't know if this would help in your case or not, but I could tell you 
which lines of code to change to turn on --debug by default.

On 01/06/16 14:19, Izaak Beekman wrote:
>
> yes its a bit tricky to get the backtrace since, at the moment it 
> pulls in a released version of JSON-Fortran, which has no 
> easy/convenient means of adding the --debug flag while attempting a 
> Homebrew build... that's what I'm working on now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/139#issuecomment-222989524>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe/AHxJPeNMluqOAZzzLWCxj_mRoCtZYgBcks5qHYb3gaJpZM4IqwMY>.
>

-- 
Chris MacMackin
cmacmackin.github.io <http://cmacmackin.github.io>

**Comment by zbeekman** [comment]
Interesting idea... yes I might be able to hack the homebrew installed FORD that way.

The alternative for me right now is trying to add patches to homebrew formulae for JSON-Fortran (to patch the JSON-Fortran CMakeLists.txt) or to submit some proposed changes (always include --debug) back to JSON-Fortran, wait for that to be merged, then perform a `--HEAD` homebrew build of JSON-Fortran

**Comment by cmacmackin** [comment]
Changing [line 173](https://github.com/cmacmackin/ford/blob/master/ford/__init__.py#L173) in the __init__.py file so that it is set to `True` should do it. Alternatively, debugging can be set in the project file. I assume that would require you to submit a new pull request or something, though.

**Comment by zbeekman** [comment]
Here is the traceback:

```
Traceback (most recent call last):
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/output.py", line 91, in __init__
    self.docs.append(TypePage(data,project,item))
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/output.py", line 186, in __init__
    self.html = self.render(data,proj,obj)
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/output.py", line 307, in render
    return template.render(data,dtype=obj,project=proj)
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/vendor/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/vendor/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/templates/type_page.html", line 1, in top-level template code
    {% extends "base.html" %}
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/templates/base.html", line 99, in top-level template code
    {% block body %}
  File "/usr/local/homebrew/Cellar/ford/4.6.0/libexec/lib/python2.7/site-packages/ford/templates/type_page.html", line 72, in block "body"
    {{ macros.bound_info(bp) }}
```

I'm also going to include the formula's "resource" blocks for getting the dependent python packages, as well as the dependencies from the formula and how/where everything is installed:

```ruby
  depends_on "graphviz"
  depends_on :python if MacOS.version <= :snow_leopard

  resource "beautifulsoup4" do
    url "https://pypi.python.org/packages/26/79/ef9a8bcbec5abc4c618a80737b44b56f1cb393b40238574078c5002b97ce/beautifulsoup4-4.4.1.tar.gz"
    sha256 "87d4013d0625d4789a4f56b8d79a04d5ce6db1152bb65f1d39744f7709a366b4"
  end

  resource "graphviz" do
    url "https://pypi.python.org/packages/3d/6d/406cec4d782d3cd6cb02d90bb17fbd364cab4a2a96d8ad0b5ccb46fd7442/graphviz-0.4.10.zip"
    sha256 "61e9f7126f5efdd11fb9269d4622277fbf8ed92046b73f3e78529e3be6a95f15"
  end

  resource "Jinja2" do
    url "https://pypi.python.org/packages/f2/2f/0b98b06a345a761bec91a079ccae392d282690c2d8272e708f4d10829e22/Jinja2-2.8.tar.gz"
    sha256 "bc1ff2ff88dbfacefde4ddde471d1417d3b304e8df103a7a9437d47269201bf4"
  end

  resource "Markdown" do
    url "https://pypi.python.org/packages/9b/53/4492f2888408a2462fd7f364028b6c708f3ecaa52a028587d7dd729f40b4/Markdown-2.6.6.tar.gz"
    sha256 "9a292bb40d6d29abac8024887bcfc1159d7a32dc1d6f1f6e8d6d8e293666c504"
  end

  resource "markdown-include" do
    url "https://pypi.python.org/packages/ef/44/eb6e9b4fa1110b719abb876c9b6dd8b46af886a94536ec4e9117fe5e7b97/markdown-include-0.5.1.tar.gz"
    sha256 "72a45461b589489a088753893bc95c5fa5909936186485f4ed55caa57d10250f"
  end

  resource "MarkupSafe" do
    url "https://pypi.python.org/packages/c0/41/bae1254e0396c0cc8cf1751cb7d9afc90a602353695af5952530482c963f/MarkupSafe-0.23.tar.gz"
    sha256 "a4ec1aff59b95a14b45eb2e23761a0179e98319da5a7eb76b56ea8cdc7b871c3"
  end

  resource "Pygments" do
    url "https://pypi.python.org/packages/b8/67/ab177979be1c81bc99c8d0592ef22d547e70bb4c6815c383286ed5dec504/Pygments-2.1.3.tar.gz"
    sha256 "88e4c8a91b2af5962bfa5ea2447ec6dd357018e86e94c7d14bd8cacbc5b55d81"
  end

  resource "toposort" do
    url "https://pypi.python.org/packages/f6/f7/875e23067652488ae40603336fdd63510a1e1853672b5b829a78452fd31c/toposort-1.4.tar.gz"
    sha256 "c190b9d9a9e53ae2835f4d524130147af601fbd63677d19381c65067a80fa903"
  end

  def install
    ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python2.7/site-packages"
    %w[beautifulsoup4 graphviz Jinja2 Markdown markdown-include MarkupSafe Pygments toposort].each do |r|
      resource(r).stage do
        system "python", *Language::Python.setup_install_args(libexec/"vendor")
      end
    end

    ENV.prepend_create_path "PYTHONPATH", libexec/"lib/python2.7/site-packages"
    system "python", *Language::Python.setup_install_args(libexec)

    bin.install Dir[libexec/"bin/*"]
    bin.env_script_all_files(libexec/"bin", :PYTHONPATH => ENV["PYTHONPATH"])
  end
```

**Comment by cmacmackin** [comment]
There's a small chance that the new version I've just released could resolve this problem.

**Comment by zbeekman** [comment]
OK, here's a backtrace for you: 

```
Creating HTML documentation...
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/output.py", line 91, in __init__
    self.docs.append(TypePage(data,project,item))
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/output.py", line 186, in __init__
    self.html = self.render(data,proj,obj)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/output.py", line 307, in render
    return template.render(data,dtype=obj,project=proj)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/type_page.html", line 1, in top-level template code
    {% extends "base.html" %}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/base.html", line 99, in top-level template code
    {% block body %}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/type_page.html", line 72, in block "body"
    {{ macros.bound_info(bp) }}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/macros.html", line 440, in template
    {{ proc_summary(bind.bindings[0]) }}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/macros.html", line 378, in template
    {% if small %}<h5>{% else %}<h3>{% endif %}{{ proc_line(proc,proto=proto) }}{% if proto %}<small>Prototype</small>{% endif %}{% if small %}</h5>{% else %}</h3>{% endif %}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/ford/templates/macros.html", line 511, in template
    {% if proc.meta['deprecated'] and proc.meta['deprecated'].lower() == 'true' %}
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/jinja2/environment.py", line 389, in getitem
    return obj[argument]
UndefinedError: 'str object' has no attribute 'meta'
```

See https://travis-ci.org/zbeekman/json-fortran/jobs/135456022#L4492-L4519 for full details, and/or clone JSON-Fortran and `./build.sh --enable-unicode`

There are a couple of suspicious "unused variable warnings" when enabling Unicode that I am looking into further, to make sure that it's not buggy Fortran causing the problem... however the comprehensive unit test framework doesn't seem to have an issue, so I'm doubtful that there's a bug in there.

**Comment by cmacmackin** [comment]
Oh no... I've seen `'str object' has no attribute 'meta'` before. That would probably indicate that that FORD is failing to match up a type bound procedure to its implementation. I thought I'd caught all of those a few releases ago. However, given that you can't reproduce this bug outside of the homebrew build environment, I wouldn't even know where to begin trying to fix it. Are you absolutely certain that an up-to-date version of FORD is being used? This issue sounds distinctly like you're using a more recent version of FORD outside of the build environment than in.

**Comment by zbeekman** [comment]
Hi Crhis,
Thanks so much for taking the time to look into this, I really appreciate it.

> Oh no... I've seen 'str object' has no attribute 'meta' before. That would probably indicate that that FORD is failing to match up a type bound procedure to its implementation. I thought I'd caught all of those a few releases ago.

I think there is a regression introduced after FORD 4.5.4, because [everything was working then](https://travis-ci.org/jacobwilliams/json-fortran/jobs/128857988#L482).

> However, given that you can't reproduce this bug outside of the homebrew build environment, I wouldn't even know where to begin trying to fix it. Are you absolutely certain that an up-to-date version of FORD is being used?

Actually, I am able to reproduce this outside of Homebrew now... I was doing something dumb and not passing the correct defines/flags to the build system outside of Homebrew... Take a look here, where I [am using FORD 4.6.1](https://travis-ci.org/zbeekman/json-fortran/jobs/135456022#L470) and the only changes since the other Travis-CI build I linked to are adding `--debug` to FORD and some trivial shell script cleanups.

**Comment by zbeekman** [comment]
Just a quick note: There is some tricky define stuff happening to invoke wrappers when UCS4 support is enabled.

**Comment by cmacmackin** [comment]
Turns out this was a side effect of the switch from calling `gfortran` for the preprocessor to `cpp`. It meant that `__GFORTRAN__` was no longer defined, which made JSON-fortran try to use a concatenation technique which still didn't work with CPP running in traditional mode. FORD will now pass in the `__GFORTRAN__` macro when it defaults to `cpp`.

**Comment by zbeekman** [comment]
Ohhhhh.... very interesting! You are a rockstar for looking into this and finding the issue! I can't thank you enough!

P.S. you should setup gratipay or gittip. I would definitely contribute a little bit to support your work on FORD.

**Comment by szaghi** [comment]
@zbeekman @cmacmackin I agree with Zaak! Chris rocks!

**Comment by cmacmackin** [comment]
When you give me a version that works and a version that doesn't, it isn't too hard to do a bisection search to find the commit that was the problem.

**Comment by zbeekman** [comment]
yes, git bisect is an amazing bug killing tool!

**Comment by zbeekman** [comment]
OK, it seems that the other issue is that on OS X I think cpp defaults to clang... That's definitely the cause of the strange Homebrew behavior, because in the Homebrew build sandbox cpp is confirmed to be clang:

```
cpp --version
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```

So it seems that JSON-Fortran needs to be clever and ensure that it specifies a good preprocessor (`gfortran -E` when compiling with gfortran, `fpp` with Intel, etc.)

**Comment by cmacmackin** [comment]
Presumably you can edit the project file using `sed` to specify the 
preprocessor to be compiler dependent. Alternatively, the preprocessor 
could be set to be the contents of an include file and then you could 
just echo the correct preprocessor into the include file? In any case, 
this is beyond the scope of FORD to deal with. Perhaps you should just 
use Linux rather than a Mac ;)

On 16-06-11 08:40 PM, Izaak Beekman wrote:
>
> OK, it seems that the other issue is that on OS X I think cpp defaults 
> to clang... That's definitely the cause of the strange Homebrew 
> behavior, because in the Homebrew build sandbox cpp is confirmed to be 
> clang:
>
> |cpp --version Apple LLVM version 7.0.2 (clang-700.1.81) Target: 
> x86_64-apple-darwin14.5.0 Thread model: posix |
>
> So it seems that JSON-Fortran needs to be clever and ensure that it 
> specifies a good preprocessor (|gfortran -E| when compiling with 
> gfortran, |fpp| with Intel, etc.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/cmacmackin/ford/issues/139#issuecomment-225389656>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe/AHxJPdywoBE19m0PyT6yJCs68XRi_sJtks5qKw8ygaJpZM4IqwMY>.
>

-- 
Chris MacMackin
cmacmackin.github.io <http://cmacmackin.github.io>

**Comment by zbeekman** [comment]
Yes, I agree that this is not really FORD's problem. However, a run time option (i.e. flag to FORD) to specify the preprocessor would be ideal. Because there is no standardization, and any given project may allow preprocessing with whatever the default tool is for the accompanying compiler. Therefore, being able to override the default preprocessor from the command line is, IMO, a necessary feature for FORD...

**Comment by jacobwilliams** [comment]
Yes, being able to override what's in the project file via a command line flag was what I was getting at over [here](https://github.com/jacobwilliams/json-fortran/issues/205).

2016 will be the year of Linux...and Fortran...on the desktop! :)
